src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs

```diff
@@ -1297,6 +1297,10 @@ public void Dispose()
     /// </summary>
     internal class PipelineStopper
     {
+        private readonly SemaphoreSlim _criticalRegionSemaphore = new SemaphoreSlim(1, 1);
```

Comment:
I think some things that should be added here in comment form are:

- What the expected scenario and thread semantics are for the critical section we're protecting
- Why `_syncRoot` isn't used or involved for this

Reading through the code it looks like the following is implicit:

- Only the first call to `EnterCriticalSection` will actually acquire the semaphore, but execution proceeds irrespective -- so another caller can "enter" the critical section before the semaphore has been released
- `AcquirePipelineStopPermission` doesn't really acquire permission, so much as wait for the critical section to be free, without ensuring that something else doesn't enter it afterward (I would call it `AwaitPipelineFree`). I assume there's a structural reason that makes this circumstance impossible, like that before we `AcquirePipelineStopPermission()` the caller sets some state that prevents the caller of `EnterCriticalSection` from calling it, or that the caller of `AcquirePipelineStopPermission()` only ever fires after all callers of `EnterCriticalSection` have been fired for a given `PipelineStopper`
- `EnterCriticalSection` also doesn't check that `AcquirePipelineStopPermission` hasn't been called in some way, so that reinforces my suspicion about the above
- Depending on the threading scenario, this looks a lot like an existing .NET lock implementation to me:
    - If we only ever expect `EnterCriticalSection` to be called by the same thread and are trying to make it work with recursive calls, we've reinvented a reentrant lock, which is provided by [Monitor](https://docs.microsoft.com/en-us/dotnet/api/system.threading.monitor?view=netcore-3.1) (i.e. `lock (_obj) { /* do stuff */ }`)
    - If we expect multiple threads to call `EnterCriticalSection` (and it somehow makes sense that they should be in the critical section concurrently), we've reinvented [ReaderWriterLock](https://docs.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlock?view=netcore-3.1)

I say all this because:

- We should be doing everything we possibly can to not reinvent lock constructions purely from a correctness perspective. Locking is just so hard to get right.
- .NET's native locks are almost certainly going to be faster

Reply:
(FWIW I've written some custom locking logic particularly in PSES, so this might strike as a tad hypocritical. But (1) we needed locks that worked reentrantly with async, (2) it's a hack and needs to be removed, and (3) every day that passes with the hack in place makes me feel worse about it and it definitely shouldn't be held up as a positive example)

Reply:
I'll get a more thorough answer once I get back to my other computer where I've been working on this... 

The main difficulty is that all the current usages here for `_syncObject` use `lock` statements which won't work for how it's currently framed. I think `Monitor` does have the ability to do enter/exit without the actual `lock` keyword, but in discussion with @SeeminglyScience he pointed out that in itself can be a bit of an antipattern.

I will freely admit this is uncharted territory for me, so if there's a better way to go about it and still get the result we're looking for, I'm all for it. 😁

Reply:
Yeah, I think this is one of those many eyes issues, where we all just need to stare it down.

The most important part is just understanding what the lock is for, who we expect the callers to be and how we anticipate they will behave. Basically what are the conceptual threads converging on this lock and why?

Reply:
Basically the _expected_ threads as I understand it to be interacting here are essentially just the thread that invokes `Stop()`/`StopProcessing()` and the thread running the current pipeline. If there are other threads getting involved there, I'm not currently aware of that being a thing.

If `cleanup{}` gets invoked _because_ of `StopProcessing()`, then all is fine, because further exceptions/cancel requests are automatically blocked off since `StopProcessing()` has already gone through once.

It's when `StopProcessing()` gets invoked _during_ `cleanup{}` that things get a bit complicated. I went through a few iterations of trying to make that work, but so far the semaphore pattern here is the only way I've managed to get it to finish the `cleanup{}` block before cancelling the pipeline. I suspect manually using Monitor would be possible here as well, but I'm not sure if that will come with additional caveats.

Reply:
Just realized I'm not sure I made this point super clear, so here goes: part of the reason we can't use `Monitor` methods here to lock based on the _syncObject directly is that `Monitor` doesn't really have a quick and easy "is something (possibly another thread) already locking this", which is what the CriticalSection semantics are for. This is important so that we can have the PipelineStopper reference that boolean property so that it can temporarily pretend that the pipeline is _not_ stopping and allow the `cleanup{}` code to run.

Without the ability to easily determine we're currently in a critical section of code, we don't have a reference point to be able to temporarily pause the halting semantics properly in a way that also properly handles potentially nested pipelines / cleanup{} blocks / etc. 

As I said above, though, if y'all have less manual solutions for this, I'm all ears!